server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -171,16 +172,40 @@ private static Query unwrap(Query query) {
     }
 
     /**
-     * Build a predicate that the "compatible" implementation of the
-     * {@link FiltersAggregator} will use to figure out if the filter matches.
-     * <p>
-     * Consumers of this method will always call it with non-negative,
-     * increasing {@code int}s. A sequence like {@code 0, 1, 7, 8, 10} is fine.
-     * It won't call with {@code 0, 1, 0} or {@code -1, 0, 1}.
+     * Returns the {@link Scorer} that the "compatible" implementation of the {@link FiltersAggregator} will use
+     * to get an iterator over the docs matching the filter. The scorer is optimized for random access, since
+     * it will be skipping documents that don't match the main query or other filters.
+     * If the passed context contains no scorer, it returns a dummy scorer that matches no docs.
      */
-    @SuppressWarnings("resource")  // Closing the reader is someone else's problem
-    IntPredicate matchingDocIds(LeafReaderContext ctx) throws IOException {
-        return Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), weight().scorerSupplier(ctx))::get;
+    Scorer randomAccessScorer(LeafReaderContext ctx) throws IOException {
+        Weight weight = weight();
+        ScorerSupplier scorerSupplier = weight.scorerSupplier(ctx);
+        if (scorerSupplier == null) {
+            return new Scorer(weight) {
+                @Override
+                public DocIdSetIterator iterator() {
+                    return DocIdSetIterator.empty();
+                }
+
+                @Override
+                public float getMaxScore(int upTo) throws IOException {
+                    return 0;
+                }
+
+                @Override
+                public float score() throws IOException {
+                    return 0;
+                }
+
+                @Override
+                public int docID() {
+                    return DocIdSetIterator.NO_MORE_DOCS;
+                }
+            };
+        }
```

Comment:
Lucene tends to return `null` scorers when there are no matches on a segment, I wonder if we should use a similar contract here for consistency.

Reply:
Done.